sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineRuleTest.java

```diff
@@ -0,0 +1,164 @@
+ */
+@RunWith(JUnit4.class)
+public class TestPipelineRuleTest implements Serializable {
+
+  private static class DummyRunner extends PipelineRunner<PipelineResult> {
+
+    @SuppressWarnings("unused") // used by reflection
+    public static DummyRunner fromOptions(final PipelineOptions opts) {
+      return new DummyRunner();
+    }
+
+    @Override
+    public PipelineResult run(final Pipeline pipeline) {
+      return new PipelineResult() {
+
+        @Override
+        public State getState() {
+          return null;
+        }
+
+        @Override
+        public State cancel() throws IOException {
+          return null;
+        }
+
+        @Override
+        public State waitUntilFinish(final Duration duration) {
+          return null;
+        }
+
+        @Override
+        public State waitUntilFinish() {
+          return null;
+        }
+
+        @Override
+        public <T> AggregatorValues<T> getAggregatorValues(final Aggregator<?, T> aggregator)
+            throws AggregatorRetrievalException {
+          return null;
+        }
+
+        @Override
+        public MetricResults metrics() {
+          return null;
+        }
+      };
+    }
+  }
+
+  private static final List<String> WORDS = Collections.singletonList("hi there");
+  private static final String EXPECTED = "expected";
+
+  private final transient TestPipeline pipeline =
+      TestPipeline.fromOptions(pipelineOptions()).enableStrictPAssert(true);
+
+  private final transient ExpectedException exception = ExpectedException.none();
+
+  @Rule
+  public transient RuleChain chain = RuleChain.outerRule(exception).around(pipeline);
+
+  private static PipelineOptions pipelineOptions() {
+    final PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+    pipelineOptions.setRunner(DummyRunner.class);
+    return pipelineOptions;
+  }
+
+  private PCollection<String> pCollection() {
+    return
+        pipeline
+            .apply(Create.of(WORDS).withCoder(StringUtf8Coder.of()))
+            .apply(MapElements.via(new SimpleFunction<String, String>() {
+
+              @Override
+              public String apply(final String input) {
+                return EXPECTED;
+              }
+            }));
+  }
+
+  @Test
+  public void testPipelineRunMissing() throws Throwable {
+    exception.expect(TestPipeline.PipelineRunMissingException.class);
+    PAssert.that(pCollection()).containsInAnyOrder(EXPECTED);
+  }
+
+  @Test
+  public void testPipelineHasAbandonedPTransform() throws Throwable {
+    exception.expect(TestPipeline.AbandonedPTransformException.class);
+    PAssert.that(pCollection()).containsInAnyOrder(EXPECTED);
+    pipeline.run().waitUntilFinish();
+    PAssert.that(pCollection()).containsInAnyOrder(EXPECTED);
+  }
+
+  @Test
+  public void testPipelinePAssertMissing() throws Throwable {
+    exception.expect(TestPipeline.PAssertMissingException.class);
+    pipeline.run().waitUntilFinish();
```

Comment:
This one I think should not be an error by default. I am not actually sure when it would be.

If there is a `PAssert` left dangling, as in the intuitive-but-broken

```
   PCollection<Foo> collection = ...
   pipeline.run();
   PAssert.that(collection)...
```

it will be caught by the general "abandoned `PTransform`" logic right?